### ISSUE TITLE
Cross validation Spider / Aragog

### ISSUE THREAD

**Issue by lsoucasse** [FormingWorlds/PROTEUS#309]
We have now Aragog running, using look up table and producing reasonable output.

But we now need a comparison with Spider to validate the implementation, see first results from @planetmariana [here](https://github.com/FormingWorlds/PROTEUS/pull/306#issuecomment-2535521679).

**Comment by planetmariana** [comment]
Hi @timlichtenberg @lsoucasse @nichollsh,

I run the exact same simulation as [here](https://github.com/FormingWorlds/PROTEUS/pull/306#issuecomment-2535521679), but with interior model=SPIDER. With those conditions the planet reached the global energy balance. Besides the time-stepping factor mentioned [#193](https://github.com/FormingWorlds/PROTEUS/issues/193), the rheological front when using Aragog starts above the CMB, so I'm a bit confused about it, as we are using the same melting curves, just in the P-T space. On the other side, both of them finish around the same time. We can discuss further about this, in the meantime I will check from the Aragog which parameter is causing the behaviour in the first 80 GPa.



![plot_interior](https://github.com/user-attachments/assets/7a789993-228d-4ee8-a44f-bb7c29a0f6c5)
![plot_global_log](https://github.com/user-attachments/assets/dc645747-0d9f-413a-bf01-febfb62add73)

**Comment by timlichtenberg** [comment]
I copy the images from the Aragog run from https://github.com/FormingWorlds/PROTEUS/pull/306#issuecomment-2535521679 to here in this thread again.
![image](https://github.com/user-attachments/assets/83f86ea3-b5d6-4cad-8e45-9468b7697f65)
![image](https://github.com/user-attachments/assets/c939563f-a94a-4e9a-8b6f-b4de59ecf628)

It looks to me as if the Aragog simulation is not initialised on an adiabat. From 80 GPa downwards the temperature profile is isothermal already at the first timestep. We need to understand the reason for this, possibly the initialisation is not yet working correctly?

**Comment by lsoucasse** [comment]
Could you send us the corresponding Aragog input file @planetmariana ?

**Comment by planetmariana** [comment]
@lsoucasse is the input file Aragog.toml in main, I just change the interior model as SPIDER or Aragog, using both with JANUS. Both simulations took around 12hr, so for now I'm checking from Aragog which parameter can be changed. As discussed before @timlichtenberg  if I change rheological_transition_melt_fraction there's a difference, but I don't know how unphysical that can be. See below:


**Output from Aragog with rheological_transition_melt_fraction=0.4** :
![mixed_phase](https://github.com/user-attachments/assets/add59456-5370-430c-b1ee-10f90499ee70)

**Output from Aragog with rheological_transition_melt_fraction=0.2**
![rheological_0 2](https://github.com/user-attachments/assets/9b1b1b94-de94-4c64-bf51-90a1beeedcfa)

**Comment by timlichtenberg** [comment]
0.4 is good and can stay. The latter is output from only running Aragog directly, and the former is using PROTEUS to execute Aragog, right? In the latter plot (just Aragog) the temperature profile looks adiabatic, it's only that the initial temperature is not high enough to generate a fully molten planet. This could mean that something is not properly communicated from PROTEUS to Aragog.

**Comment by planetmariana** [comment]
Yes, these last ones are outputs directly from Aragog. Ok, I see the problem. I will check then outputs/inputs from both sides to see where's the difference.

**Comment by timlichtenberg** [comment]
This doesn not look too bad for me, nice! :) First, the system is now cooling, check! This is different than [in the linked thread](https://github.com/FormingWorlds/PROTEUS/issues/309#issuecomment-2541154946), in there the system was initially isothermal and then heated up at the bottom and cooled at the top. The behaviour of this plot here is quite different (which is good)! SPIDER and Aragog follow a somewhat different (but both overall reasonable) cooling path and the thermal gradient of Aragog seems to follow a somewhat different behaviour, which _could_ be correct given the convective fluxes. 

We cannot quite tell, however. First, you are plotting different init time steps it seems to me (after 1 yr for SPIDER, 0 yr for Aragog). This means we cannot compare the actual convective flux. Then the cooling path is hard to follow, as we only have 2 time steps overall.

Can you please add to the plots:
- Time steps 0 yr and 1 yr for both, so we can compare the same times.
- About 5-10 time steps inbetween (if they exist). These should be all differently colored (in shades of a colormap), so we can distinguish them. Also please make sure that all time steps have the same color in both plots.

**Comment by planetmariana** [comment]
@timlichtenberg Sure ! Discussing with @lsoucasse I was wondering why Aragog take the initial temperature field and then warms to consequently gets cooler, it's following the melting curves correctly, but still a bit curious. Also, it's visible the difference in order of magnitud still on the convective heat flux as it evolves. And, the melt fraction at the lower boundary has a jump but I think that can be related to the limits I set to all the lookup tables (up to 135GPa). In any case, we realised that the problem with the liquidus being set before the pressure it's only affecting the plotting part and not the evaluation of other properties. 

**SPIDER**

![Image](https://github.com/user-attachments/assets/292d5bef-7c0b-4007-9550-5bb4d89c2710)

**Aragog**
![Image](https://github.com/user-attachments/assets/d12248ac-815f-41a5-b304-8776135274cb)

**Comment by timlichtenberg** [comment]
Thanks! Can you attached the used config files for these runs please?

**Comment by planetmariana** [comment]
Sure! 

[abe_mixed_lookup.csv](https://github.com/user-attachments/files/18619339/abe_mixed_lookup.csv)
[temperature_field.csv](https://github.com/user-attachments/files/18619331/temperature_field.csv)
[temperature_field_iso.csv](https://github.com/user-attachments/files/18619353/temperature_field_iso.csv)

**Comment by planetmariana** [comment]
Hi @timlichtenberg @lsoucasse,


I did the comparison between Aragog and SPIDER with:

- Upper boundary condition (Grey atmosphere with emmisivity=1 and Teqm=273).

- 220 nodes.

I produce this two movies to see the evolution more in detail but overall they seem to behave similarly. Although, right now I'm running new simulations with: core cooling condition and fixed heat flux at the upper boundary and higher resolution but it's taking a bit long, so I will try to run them in the cluster.

https://github.com/user-attachments/assets/114787c6-3075-47b9-b872-fad72fe0b3f2
https://github.com/user-attachments/assets/9f9761a4-6c64-41a2-b84a-e0d1b4ec2e1e

**Comment by timlichtenberg** [comment]
Awesome, thanks for the update! After our discussion yesterday, we settled on now doing the comparison for convective- and conductive-flux only with Aragog and SPIDER to see how the codes compare with only the exact same flux terms operating.

In addition, we are trying to smoothen the usage of Aragog within PROTEUS; for that we want to run PROTEUS with default settings (as discussed yesterday) and compare it by simply switching out Aragog vs. SPIDER, with settings as close as possible to each other.

**Comment by lsoucasse** [comment]
I have tried to run the default settings with Aragog (and Janus) but it is extremely slow.
It gets stuck at iteration 6 which is the first significant time step (300 years) but did not completed it after 2 hours...
Do you experience such long running time Mariana?

**Comment by planetmariana** [comment]
Hi @lsoucasse , yes when using Aragog+JANUS it can take up to 8hr with default settings.

**Comment by timlichtenberg** [comment]
Can you do a comparison with AGNI?

**Comment by lsoucasse** [comment]
> Can you do a comparison with AGNI?

I tried with AGNI but it fails to converge and provide a solution with default settings. Here is the log.

[failed-coupled-agni.log](https://github.com/user-attachments/files/18798986/failed-coupled-agni.log)

**Comment by nichollsh** [comment]
AGNI also seems to be performing very slowly for you. In one of the PROTEUS steps, it only manages a single iteration before timing-out.

**Comment by nichollsh** [comment]
Which points to a SOCRATES issue. I also noticed that sometimes SOCRATES was returning garbage values in the GitHub actions. I don't think this is a fault of gfortran, but it points to a problem with the spectral files. My suspicion is that OSF could be providing corrupted data...

**Comment by timlichtenberg** [comment]
I notice that the volatile settings are somewhat extreme in the .toml file. I cannot find earth-demo.toml in current `main`. Is that the same as `default.toml`?

**Comment by timlichtenberg** [comment]
It's probably okay to stop at mantle solidification for the demo, but that should be followed up I guess.

**Comment by lsoucasse** [comment]
Now Aragog has the feature to initiate the temperature profile with an adiabat.

Still, when running Proteus using branch [ls/aragog-adiabat](https://github.com/FormingWorlds/PROTEUS/tree/ls/aragog-adiabat) and the default input (replacing Spider by Aragog) the computation gets stuck after the init steps.

Could you see if you can reproduce this @planetmariana 

[proteus_00.log](https://github.com/user-attachments/files/19316067/proteus_00.log)

**Comment by planetmariana** [comment]
Hi @lsoucasse this is my aragog.py file and the initial temperature file that I'm using to run PROTEUS.

```
# Aragog interior module
from __future__ import annotations

import glob
import logging
import os
from pathlib import Path
from typing import TYPE_CHECKING

import netCDF4 as nc
import numpy as np
import pandas as pd
import platformdirs
from aragog import Output, Solver, aragog_file_logger
from aragog.parser import (
    Parameters,
    _BoundaryConditionsParameters,
    _EnergyParameters,
    _InitialConditionParameters,
    _MeshParameters,
    _PhaseMixedParameters,
    _PhaseParameters,
    _Radionuclide,
    _ScalingsParameters,
    _SolverParameters,
)

from proteus.interior.common import Interior_t
from proteus.interior.timestep import next_step
from proteus.utils.constants import R_earth, radnuc_data, secs_per_year

if TYPE_CHECKING:
    from proteus.config import Config

aragog_solver = None
log = logging.getLogger("fwl."+__name__)

FWL_DATA_DIR = Path(os.environ.get('FWL_DATA', platformdirs.user_data_dir('fwl_data')))

# Run the Aragog interior module
def RunAragog(config:Config, dirs:dict,
                hf_row:dict, hf_all:pd.DataFrame, interior_o:Interior_t):

    global aragog_solver

    # Setup Aragog logger
    aragog_file_logger(console_level = logging.WARNING,
                       file_level = logging.getLevelName(config.params.out.logging),
                       log_dir = dirs["output"])

    # Compute time step
    if interior_o.ic==1:
        dt = 0.0
        aragog_solver = None
    else:
        step_sf = 1.0 # dt scale factor
        dt = next_step(config, dirs, hf_row, hf_all, step_sf)

    # Setup Aragog parameters from options at first iteration
    if (aragog_solver is None):
        SetupAragogSolver(config, hf_row, interior_o)
        # Update state from stored data if resuming a simulation
        if config.params.resume:
            UpdateAragogSolver(dt, hf_row, interior_o, output_dir=dirs["output"])
    # Update varying parameters in ongoing simulation
    else:
        UpdateAragogSolver(dt, hf_row, interior_o)

    # Run Aragog solver
    aragog_solver.initialize()
    aragog_solver.solve()

    # Get Aragog output
    output = GetAragogOutput(hf_row, interior_o)
    sim_time = aragog_solver.parameters.solver.end_time

    # Write output to a file
    WriteAragogOutput(dirs["output"],sim_time)

    return sim_time, output


def SetupAragogSolver(config:Config, hf_row:dict, interior_o:Interior_t):

    global aragog_solver

    scalings = _ScalingsParameters(
            radius = R_earth, # scaling radius [m]
            temperature = 4000, # scaling temperature [K]
            density = 4000, # scaling density
            time = secs_per_year, # scaling time [sec]
            )

    solver = _SolverParameters(
            start_time = 0,
            end_time = 0,
            atol = config.interior.aragog.tolerance,
            rtol = config.interior.aragog.tolerance,
            )

    boundary_conditions = _BoundaryConditionsParameters(
            outer_boundary_condition = 4, # 4 = prescribed heat flux
            outer_boundary_value = hf_row["F_atm"], # first guess surface heat flux [W/m2]
            inner_boundary_condition = 2, # 1 = core cooling model, 3 = prescribed temperature
            inner_boundary_value = 0.11, # heat flux from maximum amount at the Earth's CMB over the surface area of the radius. core temperature [K], if inner_boundary_condition = 3
            emissivity = 1, # only used in gray body BC, outer_boundary_condition = 1
            equilibrium_temperature = hf_row["T_eqm"], # only used in gray body BC, outer_boundary_condition = 1
            core_density = config.struct.core_density, # used if inner_boundary_condition = 1
            core_heat_capacity = 880, # used if inner_boundary_condition = 1
            )

    mesh = _MeshParameters(
            outer_radius = hf_row["R_int"], # planet radius [m]
            inner_radius = config.struct.corefrac * hf_row["R_int"], # core radius [m]
            number_of_nodes = config.interior.aragog.num_levels, # basic nodes
            mixing_length_profile = "constant",
            surface_density = 4090, # AdamsWilliamsonEOS parameter [kg/m3]
            gravitational_acceleration = hf_row["gravity"], # [m/s-2]
            adiabatic_bulk_modulus = config.interior.bulk_modulus, # AW-EOS parameter [Pa]
            )

    energy = _EnergyParameters(
            conduction = True,
            convection = True,
            gravitational_separation = False,
            mixing = False,
            radionuclides = config.interior.radiogenic_heat,
            tidal = config.interior.tidal_heat,
            # tidal_array = interior_o.tides
            )

    # Get look up data directory, will be configurable in the future
    LOOK_UP_DIR = FWL_DATA_DIR / "interior_lookup_tables/1TPa-dK09-elec-free/MgSiO3_Wolf_Bower_2018/"

    initial_condition = _InitialConditionParameters(
            # surface_temperature = 3200, # initial top temperature (K)
            # basal_temperature = 5039, # initial bottom temperature (K)
            init_file= LOOK_UP_DIR / "mantle_adiabat_results.dat"
            )


    phase_liquid = _PhaseParameters(
            density = LOOK_UP_DIR / "density_melt.dat",
            viscosity = 1E2,
            heat_capacity = LOOK_UP_DIR / "heat_capacity_melt.dat",
            melt_fraction = 1,
            thermal_conductivity = 4,
            #thermal_expansivity = 1.0E-5,
            thermal_expansivity = LOOK_UP_DIR / "thermal_exp_melt.dat",
            )

    phase_solid = _PhaseParameters(
            density = LOOK_UP_DIR / "density_solid.dat",
            viscosity = 1E21,
            heat_capacity = LOOK_UP_DIR / "heat_capacity_solid.dat",
            melt_fraction = 0,
            thermal_conductivity = 4,
            #thermal_expansivity = 1.0E-5,
            thermal_expansivity = LOOK_UP_DIR / "thermal_exp_solid.dat",
            )

    phase_mixed = _PhaseMixedParameters(
            latent_heat_of_fusion = 4e6,
            rheological_transition_melt_fraction = config.interior.rheo_phi_loc,
            rheological_transition_width = config.interior.rheo_phi_wid,
            solidus = LOOK_UP_DIR / "solidus.dat",
            liquidus = LOOK_UP_DIR / "liquidus.dat",
            phase = "mixed",
            phase_transition_width = 0.1,
            grain_size = config.interior.grain_size,
            )

    radionuclides = []
    if config.interior.radiogenic_heat:
        # offset by age_ini, which converts model simulation time to the actual age
        radio_t0 = config.delivery.radio_tref - config.star.age_ini
        radio_t0 *= 1e9 # Convert Gyr to yr

        def _append_radnuc(_iso, _cnc):
            radionuclides.append(_Radionuclide(
                                    name = _iso,
                                    t0_years = radio_t0,
                                    abundance = radnuc_data[_iso]["abundance"],
                                    concentration = _cnc,
                                    heat_production = radnuc_data[_iso]["heatprod"],
                                    half_life_years = radnuc_data[_iso]["halflife"],
                                ))

        if config.delivery.radio_K > 0.0:
            _append_radnuc("k40", config.delivery.radio_K)

        if config.delivery.radio_Th > 0.0:
            _append_radnuc("th232", config.delivery.radio_Th)

        if config.delivery.radio_U > 0.0:
            _append_radnuc("u235", config.delivery.radio_U)
            _append_radnuc("u238", config.delivery.radio_U)

    param = Parameters(
            boundary_conditions = boundary_conditions,
            energy = energy,
            initial_condition = initial_condition,
            mesh = mesh,
            phase_solid = phase_solid,
            phase_liquid = phase_liquid,
            phase_mixed = phase_mixed,
            radionuclides = radionuclides,
            scalings = scalings,
            solver = solver,
            )

    aragog_solver = Solver(param)

def UpdateAragogSolver(dt:float, hf_row:dict, interior_o:Interior_t,
                            output_dir:str = None):

    # Set solver time
    # hf_row["Time"] is in yr so do not need to scale as long as scaling time is secs_per_year
    aragog_solver.parameters.solver.start_time = hf_row["Time"]
    aragog_solver.parameters.solver.end_time = hf_row["Time"] + dt

    # Get temperature field from previous run
    if output_dir is not None: # read it from output directory
        Tfield = read_last_Tfield(output_dir, hf_row["Time"])
    else: # get it from solver
        Tfield = aragog_solver.temperature_staggered[:, -1]

    # Update initial condition
    Tfield = Tfield / aragog_solver.parameters.scalings.temperature
    aragog_solver.parameters.initial_condition.from_field = True
    aragog_solver.parameters.initial_condition.init_temperature = Tfield

    # Update boundary conditions
    aragog_solver.parameters.boundary_conditions.outer_boundary_value = hf_row["F_atm"]

    # Update tidal heating within the mantle
    aragog_solver.parameters.energy.tidal_array = \
        interior_o.tides / aragog_solver.parameters.scalings.power_per_mass

def WriteAragogOutput(output_dir:str, time:float):

    aragog_output: Output = Output(aragog_solver)

    fpath = os.path.join(output_dir,"data","%d_int.nc"%time)
    aragog_output.write_at_time(fpath,-1)

def GetAragogOutput(hf_row:dict, interior_o:Interior_t):

    aragog_output: Output = Output(aragog_solver)
    aragog_output.state.update(aragog_output.solution.y, aragog_output.solution.t)
    output = {}

    output["M_mantle"] = aragog_output.mantle_mass
    output["T_magma"] = aragog_output.solution_top_temperature
    output["Phi_global"] = float(aragog_output.melt_fraction_global)
    output["RF_depth"] = float(aragog_output.rheological_front)
    output["F_int"] = aragog_output.convective_heat_flux_basic[-1,-1] # Need to be revised for consistency

    output["M_mantle_liquid"] = output["M_mantle"] * output["Phi_global"]
    output["M_mantle_solid"] = output["M_mantle"] - output["M_mantle_liquid"]

    # Calculate surface area
    radii = aragog_output.radii_km_basic * 1e3 # [m]
    area  = 4 * np.pi * radii[-1]**2 # [m^2]

    # Mass at each mesh layer
    mass_s = aragog_output.mass_staggered[:,-1] # [kg]

    # Radiogenic heating
    Hradio_s = aragog_output.heating_radio[:,-1]  # [W kg-1]
    output["F_radio"] = np.dot(Hradio_s, mass_s) / area  # [W m-2]

    # Tidal heating flux
    Htidal_s = aragog_output.heating_tidal[:,-1]  # [W kg-1]
    output["F_tidal"] = np.dot(Htidal_s, mass_s)/area

    # Store arrays
    # FIX ME - Should extract values from staggered nodes rather than cropping basic nodes.
    interior_o.phi      = np.array(aragog_output.melt_fraction_staggered[:,-1])
    interior_o.visc     = np.power(10.0, aragog_output.log10_viscosity_staggered[:,-1])
    interior_o.density  = np.array(aragog_output.density_basic[:,-1])[1:]
    interior_o.radius   = radii[:] # length N+1
    interior_o.mass     = mass_s[:]
    interior_o.temp     = np.array(aragog_output.temperature_K_staggered[:,-1])
    interior_o.pres     = np.array(aragog_output.pressure_GPa_staggered[:,-1] * 1e9)

    return output

def read_last_Tfield(output_dir:str, time:float):

    # Read Aragog output at last run
    fpath = os.path.join(output_dir,"data","%d_int.nc"%time)
    out = read_ncdf(fpath)

    # Get temperature field at basic nodes and interpolate to staggered nodes
    T_basic = out["temp_b"]
    N_basic = len(T_basic)
    T_staggered = (T_basic[0:N_basic-1] + T_basic[1:N_basic] ) / 2.0

    return T_staggered

def get_all_output_times(output_dir:str):
    files = glob.glob(output_dir+"/data/*_int.nc")
    years = [int(f.split("/")[-1].split("_int")[0]) for f in files]
    mask = np.argsort(years)

    return [years[i] for i in mask]

def read_ncdf(fpath:str):
    out = {}
    ds = nc.Dataset(fpath)

    for key in ds.variables.keys():
        out[key] = ds.variables[key][:]

    ds.close()
    return out

def read_ncdfs(output_dir:str, times:list):
    return [read_ncdf(os.path.join(output_dir, "data", "%d_int.nc"%t)) for t in times]

```


[mantle_adiabat_results.csv](https://github.com/user-attachments/files/19336204/mantle_adiabat_results.csv)

**Comment by lsoucasse** [comment]
If I remember correctly `F_int`in proteus is clipped to zero when it becomes negative.

**Comment by timlichtenberg** [comment]
Yeah that looks interesting... What are the conditions of these simulations? All heat fluxes aside from convective turned off? Indeed you would need to output the core heat flux; and also run a simulation with constant lower boundary conditions, as clearly this time-changing heat flux conditions is not yet feasible. Can I also ping @djbower here, as I feel this issue is what is holding us back to make the next step. Any suggestions on how to debug this more closely?

**Comment by djbower** [comment]
Try and isolate the problem as best as possible.  Maybe anchor the CMB to fixed temperature (isothermal) and see if the surface then avoids the spikes.  @nichollsh is correct that the spikes likely relate to convection abruptly turning off if one or several cells become sub-adiabatic.  So identifying when/where the convection criteria is no longer satisfied can be helpful.  Also check the original SPIDER C code to look for smoothing functions because the original code might smooth the fluxes.  This certainly happens across phase transitions, such as the liquidus and solidus, but I don't recall if some quantities are also smoothed in the mixed phase region.

**Comment by planetmariana** [comment]
Hi @timlichtenberg @lsoucasse @nichollsh @djbower ,

I implemented your suggestions, to isolate the problem as much as possible, setting all the values to a constant value and with the initial adiabat to be exactly the same,  as you can see in the next video. The super adiabatic gradient already from the first time-step differs quite a lot.

https://github.com/user-attachments/assets/76f081c2-5146-4945-9f20-d7e2ea7dfb61


So, I went back to the equations:

1) In Aragog:

$\nabla T_{\text{super}} = \frac{dT}{dr} - \left( \frac{dT}{dr} \right)_{\text{adiabatic}}$
 ` self._super_adiabatic_temperature_gradient = self.dTdr() - self.phase_basic.dTdrs() ` 

where:

$\left( \frac{dT}{dr} \right)_{\text{adiabatic}}= - \frac{g \cdot \alpha \cdot T}{c_p}$

and:

$\frac{dT}{dr}$  is computed through finite differences


2) In SPIDER is computed through the derivatives of the entropy field directly:

$\nabla T_{\text{super}} = \frac{T}{c_p} \cdot \frac{dS}{d\xi} \cdot \frac{d\xi}{dr}$
 ` 
super_adiabatic_gradient=eos_eval.T / eos_eval.Cp * dSdxi * dxidr
 ` 

So, I isolated the terms corresponding to dTdr and dTdrs and the resulting Super-adiabatic gradient in Aragog and this is the result:

https://github.com/user-attachments/assets/10940922-d1e8-429c-9f6d-9bd011fdf6ab


which is already 4 orders of magnitude lower than the super-adiabatic-gradient from SPIDER that you can see in the first video. As a remark both of them have the same initial conditions, prescribed heat_flux at top and bottom boundary. I also checked the smoothing of functions in the mixed phase region that SPIDER is doing as Dan suggested, and Aragog is also implementing them in the CompositePhaseEvaluator and MixedPhaseEvaluator classes with the function:

```python
def _get_composite(self, property_name: str) -> npt.NDArray:
    """Evaluates the composite property"""
    mixed_phase: npt.NDArray = getattr(self._mixed, property_name)()
    single_phase: npt.NDArray = np.empty_like(self._blending_factor)
    test = getattr(self._liquid, property_name)()

    try:
        single_phase[self._liquid_mask] = getattr(self._liquid, property_name)()[
            self._liquid_mask
        ]
    except (IndexError, TypeError):
        single_phase[self._liquid_mask] = getattr(self._liquid, property_name)()
    try:
        single_phase[self._solid_mask] = getattr(self._solid, property_name)()[
            self._solid_mask
        ]
    except (IndexError, TypeError):
        single_phase[self._solid_mask] = getattr(self._solid, property_name)()

    combined: npt.NDArray = combine_properties(
        self._blending_factor, mixed_phase, single_phase
    )

    return combined
```

and this is applied to the density, heat capacity, thermal conductivity, thermal expansivity and dTdPs. At this point, my intuition is that something is not going correctly with the computation of dTdr, I tried to smooth the temperature before computing dTdr here:

```
self._dTdr = self._evaluator.mesh.d_dr_at_basic_nodes(temperature)
```

But that means, I would need to smooth the temperature at every time step and is not computationally effective. That's my update right now, I'm at this point, if you have any advice I would really appreciate it, as I couldn't go forward after this. Thanks :)

**Comment by nichollsh** [comment]
Thanks for the detailed update, @planetmariana

Regarding this point:
> already 4 orders of magnitude lower than the super-adiabatic-gradient from SPIDER that you can see in the first video. As a remark both of them have the same initial conditions, 

It's not clear to me why the first $dT/dr$ gradient term (from finite-difference of $T(r)$) would be 4 OOM different between the two models. Surely this term is entirely defined by the initial $T(r)$ profile that you provide, and you are providing the same $T(r)$ profile to both models.

Could this 1e4 difference be reflecting some issue (either in the calculation or the plots) with the constants used to non-dimensionalise the calculations?

**Comment by djbower** [comment]
You might have to copy-paste-convert some of the original c-code in SPIDER associated with the initial condition into Aragag/Python, and then as Harrison eludes to above, focus first on exactly reproducing the temperature gradient from the initial condition.

**Comment by djbower** [comment]
I just had a brief chat with @timlichtenberg on this topic.  Here are a few points to think about:

1. I would focus on the initial condition.  Systematically compare the array values, particularly for temperature and its gradient, to determine how comparable they are.  Within the code the quantities are non-dimensional. according to the scalings that are set in the relevant configuration file.  Then compare the values in the outputs.  Note that you may have to scale the outputs by the relevant scaling to convert them to dimensional units.  In short, the simplest reason for the discrepancy would be that a dimensional scaling such as (TEMPERATURE/RADIUS) for the temperature gradient is not being applied or accounted for when making the comparison.

2. Similarly, compare other array outputs/calculations, such as the super-adiabatic temperature gradient.  Again, be careful to perform a like-for-like comparison, either the non-dimensional values (but ensure the same scaling between SPIDER and Aragog), or alternatively the physical (scaled) values.

3. The time integration for SPIDER and Aragog is not the same and could introduce some differences during the integration due to numerical noise.  Hence analyse the initial state as much as possible before performing time integrations.

4. Mass coordinates will not introduce differences of 4 orders of magnitude.

In the end, since you have two codes, you can systematically debug until you find the discrepancy.  It's a bit laborious, granted, but should be achievable to narrow down the difference.  I hope these pointers help to keep you moving.

**Comment by timlichtenberg** [comment]
Thanks Dan, we will discuss the next steps a bit further in our meeting on Thursday.

**Comment by timlichtenberg** [comment]
This boils down to overall issues:
1. Validate the equation system. This *should* be working, but we need to double check
2. Validate the translation of key input parameters from S to T formulation.

I will accordingly open two new related issues.

**Comment by planetmariana** [comment]
Based on our previous discussion here, I extracted the quantities that are involved in the computation of the convective heat flux, in both SPIDER and Aragog and created this plot where each column correspond to each quantity in the equation of Fconv and:

1st row: Unscaled quantity
2nd row: Scaling factor
3rd row: Scaled quantity

As a remark: I set a constant value for the Heat Capacity, Density and Thermal Expansivity on the lookup tables on both codes to avoid modifying the input files.

Blue curves: Aragog
Orange curves: SPIDER
 
Also, SATG means "Super Adiabatic Temperature Gradient"

<img width="1507" alt="Image" src="https://github.com/user-attachments/assets/5afe554f-8014-4272-b473-5a21122cdc79" />

I attach the full movie, although I don't think on GitHub it conserves the full quality :/

https://github.com/user-attachments/assets/5502dbed-25a0-4c3f-ab8e-df158626b7c3

From here, it's possible to see that comparing even scaled or unscaled, both SATG are still different in both codes by at least 4 OOM, affecting the convective heat flux specially on the edges which is what the plot [above in this issue](https://github.com/FormingWorlds/PROTEUS/issues/309#issuecomment-2912027143) is showing. Then on previous meeting with @lsoucasse , we agree on test a linear profile on Aragog that we could test analytically, such that:

$$T(r) = A(r - r_0) + T_0$$

$$
\frac{dT^{\text{super}}}{dr} = A - \frac{g \alpha}{c_p} \left( A(r - r_0) + T_0 \right)
$$

With  $T_0$ being the temperature at the CMB at $r_0$, in this case we took T>Tliq. In that way, if we have everything else constant therefore we would expect the SATG to be linear, which was the case. Then we said: ok, what if we try to put the linear temperature profile in SPIDER and with the same conditions comparing the scaled value, we would expect the same SATG as a validation. And for that I needed to map the linear temperature profile into the P-S space to put it on SPIDER as an initial condition. When doing that I realised that a T profile that is linear on Aragog in the P-T space was having this shape on the P-S space (I corrected from the one this morning, sorry my mistake :( ):

![Image](https://github.com/user-attachments/assets/b4619b1a-3a3f-4137-8669-d9be2a6d9830)

Anyway, when doing that, PETSc complains about the initial file, so I was wondering if the initial entropy field must have a specific format to be reader by PETSc? I already check that the field contains the same staggered nodes as the one set on the input file. 

![Image](https://github.com/user-attachments/assets/3df82036-2794-4080-bf0d-4a44aad9bbda)

This is how I'm setting the initial condition on SPIDER:

```
# initial condition
# [1] is set dS/dr to constant everywhere
# [2] is restart from file
-IC_INTERIOR 2
-ic_interior_filename /Users/marianity/Desktop/PROTEUS/SPIDER/lookup_data/1TPa-dK09-elec-free/entropy_field.m
```
where entropy_field.m looks like:

![Image](https://github.com/user-attachments/assets/4b50fb89-8359-48ee-8ba0-d83bc4b8d2ed)
@lsoucasse @timlichtenberg If I'm missing something else please add it here too. Thanks

**Comment by djbower** [comment]
Check the source code in SPIDER for the input format (I don't remember exactly the format) But from a quick glance perhaps you need a space after the first # before the number 5?

**Comment by timlichtenberg** [comment]
Thanks a lot Mariana for putting this together, Mariana! :-) This is will be very useful as we return in a bit to the Aragog-PROTEUS integration.
